### ISSUE TITLE
No way to ignore warnings

### ISSUE THREAD

**Issue by IgnacioJPickering** [rdkit/rdkit#2683]
**I can't find a way to ignore RDKit's warnings**

I'm using RDKit version 2019.03.4; I haven't been able to suppress the warnings that 
pop up when trying to convert a SMILES string to a molecule object.

I have to parse a LARGE number of smiles (~46 M strings) and all of them have hydrogen atoms without neighbors. Each time RDKit parses one of the smiles strings I get the warning

```[17:58:45] WARNING: not removing hydrogen atom without neighbors ```

Since the number of strings is so large, this results in a significant slowdown...

My current workaround is to redirect ```stderr``` to ```/dev/null```, but there surely is a better solution? I couldn't find it, and it seems that python's warnings module doesn't interact with RDKit's warnings at all.

**Comment by greglandrum** [comment]
@IgnacioJPickering: the RDKit warnings are done at the C++ level and, you are correct, they don't integrate well with the normal Python warnings module.

You can, however, switch off the messages using functionality in the `RDLogger` module:
```
In [1]: from rdkit import Chem                                                                                                                                                                   

In [2]: from rdkit import RDLogger                                                                                                                                                               

In [3]: Chem.MolFromSmiles('c1cncc1')                                                                                                                                                            
[09:26:28] Can't kekulize mol.  Unkekulized atoms: 0 1 2 3 4


In [4]: RDLogger.DisableLog('rdApp.*')                                                                                                                                                           

In [5]: Chem.MolFromSmiles('c1cncc1')                                                                                                                                                            

In [6]:  
```

**Comment by Minys233** [comment]
> @IgnacioJPickering: the RDKit warnings are done at the C++ level and, you are correct, they don't integrate well with the normal Python warnings module.
> 
> You can, however, switch off the messages using functionality in the `RDLogger` module:
> 
> ```
> In [1]: from rdkit import Chem                                                                                                                                                                   
> 
> In [2]: from rdkit import RDLogger                                                                                                                                                               
> 
> In [3]: Chem.MolFromSmiles('c1cncc1')                                                                                                                                                            
> [09:26:28] Can't kekulize mol.  Unkekulized atoms: 0 1 2 3 4
> 
> 
> In [4]: RDLogger.DisableLog('rdApp.*')                                                                                                                                                           
> 
> In [5]: Chem.MolFromSmiles('c1cncc1')                                                                                                                                                            
> 
> In [6]:  
> ```

Thank you so much! After quite a lot googling, finally solved this problem XD

**Comment by ademidun** [comment]
Thanks for this suggestion @greglandrum , this was about 90% of what I needed.

For people coming here through Google who like me have a slightly different use case:

If you only want to see a specific warning for example we wanted to disable
`WARNING: not removing hydrogen atom without neighbors` but we want to see the logs for ` Can't kekulize mol.  Unkekulized atoms: 0 1 2 3 4`

Also, full disclosure, I haven't tried this but I think it should work.

```python
RDLogger.DisableLog('rdApp.info')                                                                                                                                                       
Chem.MolFromSmiles('<long_SMILES_STRING_TO_IGNORE>')  # this line won't print                                                                                                                                                           
RDLogger.EnableLog('rdApp.info')                                                                                                                                                        
Chem.MolFromSmiles('c1cncc1') # this line will print     
```

**Comment by sorenwacker** [comment]
I am still getting lots of log messages:

[09:52:41] Tautomer enumeration stopped at 285 tautomers: max transforms reached
[09:52:42] Tautomer enumeration stopped at 716 tautomers: max transforms reached
[09:52:43] Tautomer enumeration stopped at 758 tautomers: max transforms reached
[09:52:44] Tautomer enumeration stopped at 1000 tautomers: max tautomers reached
[09:52:47] Tautomer enumeration stopped at 312 tautomers: max transforms reached
[09:52:49] Tautomer enumeration stopped at 663 tautomers: max transforms reached
[09:52:50] Tautomer enumeration stopped at 210 tautomers: max transforms reached
[09:52:51] Tautomer enumeration stopped at 1000 tautomers: max tautomers reached
[09:52:53] Tautomer enumeration stopped at 729 tautomers: max transforms reached
[09:52:53] WARNING: not removing hydrogen atom without neighbors
[09:52:53] WARNING: not removing hydrogen atom without neighbors
[09:52:53] WARNING: not removing hydrogen atom without neighbors
[09:52:54] Tautomer enumeration stopped at 143 tautomers: max transforms reached

Is there a way to suppress that.

**Comment by Vencent-Won** [comment]
Can I capture these warnings by "try&except"?

**Comment by dhibbit** [comment]
> @IgnacioJPickering: the RDKit warnings are done at the C++ level and, you are correct, they don't integrate well with the normal Python warnings module.
> 
> You can, however, switch off the messages using functionality in the `RDLogger` module:
> 
> ```
> In [1]: from rdkit import Chem                                                                                                                                                                   
> 
> In [2]: from rdkit import RDLogger                                                                                                                                                               
> 
> In [3]: Chem.MolFromSmiles('c1cncc1')                                                                                                                                                            
> [09:26:28] Can't kekulize mol.  Unkekulized atoms: 0 1 2 3 4
> 
> 
> In [4]: RDLogger.DisableLog('rdApp.*')                                                                                                                                                           
> 
> In [5]: Chem.MolFromSmiles('c1cncc1')                                                                                                                                                            
> 
> In [6]:  
> ```

This doesn't appear to work for warnings generated by rdDetermineBonds.DetermineBonds(), although it did suppress warnings from Chem.inchi.MolToInchi().

**Comment by amorehead** [comment]
@dhibbit, did you ever find a way to suppress warnings or logs coming from calls to the `rdDetermineBonds.DetermineBonds` function?

**Comment by greglandrum** [comment]
@amorehead unless you share code and example input that produces the output you are unhappy about, I have no chance to assist

**Comment by amorehead** [comment]
Hi, @greglandrum. Thanks for your quick reply. The following function is one I am using to construct `Chem.Mol` objects from (`.xyz`-like) 3D coordinate and element type input NumPy arrays. When I run this function, I see `rdDetermineBonds.DetermineBonds(mol, charge=charge)`'s verbose (I assume C++-level) output in my terminal, even though I have tried to suppress these logs using `RDLogger.DisableLog("rdApp.*")` called at the top of the script containing this function.

```python
def create_mol_from_atom_positions_and_types(
    name: str,
    atom_positions: np.ndarray,
    element_types: List[str],
    missing_atom_indices: Set[int],
    num_bond_attempts: int = 2,
    verbose: bool = False,
) -> Mol:
    """Create an RDKit molecule from a NumPy array of atom positions and a list of their element
    types.

    :param name: The name of the molecule.
    :param atom_positions: A NumPy array of shape (num_atoms, 3) containing the 3D coordinates of
        each atom.
    :param element_types: A list of element symbols for each atom in the molecule.
    :param missing_atom_indices: A set of atom indices that are missing from the atom_positions
        array.
    :param num_bond_attempts: The number of attempts to determine the bonds in the molecule.
    :param verbose: Whether to log warnings when bond determination fails.
    :return: An RDKit molecule with the specified atom positions and element types.
    """
    if len(atom_positions) != len(element_types):
        raise ValueError("The length of atom_elements and xyz_coordinates must be the same.")

    # populate an empty editable molecule

    mol = Chem.RWMol()
    mol.SetProp("_Name", name)

    for element_type in element_types:
        atom = Chem.Atom(element_type)
        mol.AddAtom(atom)

    # set 3D coordinates

    conf = Chem.Conformer(mol.GetNumAtoms())
    for i, (x, y, z) in enumerate(atom_positions):
        conf.SetAtomPosition(i, Point3D(x, y, z))

    mol.AddConformer(conf)
    Chem.SanitizeMol(mol)

    # finalize molecule by inferring bonds

    determined_bonds = False
    for i in range(num_bond_attempts):
        try:
            charge = Chem.GetFormalCharge(mol)
            rdDetermineBonds.DetermineBonds(mol, charge=charge)
            determined_bonds = True
        except Exception as e:
            if verbose:
                logger.warning(
                    f"Failed to determine bonds for the input molecule {name} due to: {e}. "
                    f"{'Retrying once more.' if i < num_bond_attempts - 1 else 'Terminating bond assignment.'}"
                )
            continue
    if not determined_bonds:
        if verbose:
            logger.warning(
                "Failed to determine bonds in the input molecule. Skipping bond assignment."
            )

    mol = Chem.RemoveHs(mol)
    Chem.SanitizeMol(mol)

    # set a property to indicate the atom positions that are missing

    mol.SetProp("missing_atom_indices", ",".join(map(str, sorted(missing_atom_indices))))

    return mol
```

**Comment by bp-kelley** [comment]
The new standard way to block logging events is to use rdBase.BlockLogs()

This is a scoped variable which has the nice feature that when it goes out of scope, the logs are unblocked.  Example

```
>>> from rdkit import rdBase
>>> from rdkit import Chem
>>> blocker = rdBase.BlockLogs()
>>> Chem.MolFromSmiles("asdf")
>>> del blocker
>>> Chem.MolFromSmiles("asdf")
[11:27:27] SMILES Parse Error: syntax error while parsing: asdf
[11:27:27] SMILES Parse Error: Failed parsing SMILES 'asdf' for input: 'asdf'
```

Hopefully this works for you.

**Comment by bp-kelley** [comment]
(i.e add blocker = rdBase.BlockLogs()) to the top of your function.

**Comment by amorehead** [comment]
Thanks, @bp-kelley! I'll give it a try

**Comment by amorehead** [comment]
@bp-kelley, sadly, it doesn't look like that fixed the issue. I'm still seeing logs like the following being printed to my terminal when I call RDKit's bond determination functions:
```
        0 7 N chg: -2  deg: 1 exp: 0 imp: 0 hyb: SP3
        1 6 C chg: 1  deg: 3 exp: 0 imp: 0 hyb: SP3
        2 6 C chg: 0  deg: 2 exp: 0 imp: 0 hyb: SP3
        3 6 C chg: 1  deg: 3 exp: 0 imp: 0 hyb: SP3
        4 8 O chg: -1  deg: 1 exp: 0 imp: 0 hyb: SP3
        5 6 C chg: -3  deg: 1 exp: 0 imp: 0 hyb: SP3
        6 6 C chg: -3  deg: 1 exp: 0 imp: 0 hyb: SP3
Bonds:
        0 1->0 order: 1
        1 2->1 order: 1
        2 3->1 order: 1
        3 4->2 order: 1
        4 5->3 order: 1
        5 6->3 order: 1
Atoms:
        0 7 N chg: -2  deg: 1 exp: 0 imp: 0 hyb: SP3
        1 6 C chg: 0  deg: 2 exp: 0 imp: 0 hyb: SP3
        2 6 C chg: 0  deg: 2 exp: 0 imp: 0 hyb: SP3
        3 8 O chg: -1  deg: 1 exp: 0 imp: 0 hyb: SP3
Bonds:
        0 1->0 order: 1
        1 2->1 order: 1
        2 3->2 order: 1
```

**Comment by bp-kelley** [comment]
Ok, this looks like a proper bug, that is coming from mol.DebugMol which isn't being sent to a log.  I'll take a look for where it might be happening.

**Comment by bp-kelley** [comment]
Fixed in https://github.com/rdkit/rdkit/pull/7690

**Comment by bp-kelley** [comment]
@amorehead please follow #7691

**Comment by amorehead** [comment]
@bp-kelley, is it possible to `pip install` the current state of the main branch to use this fix? If so, which command should I use.

**Comment by bp-kelley** [comment]
No, we don't support building from pip.  this will be pushed in the next release.  There is usually one done before the RDKit UGM.  @greglandrum might have more details.

**Comment by amorehead** [comment]
OK. This might make it even trickier, since I have to install RDKit using `conda` for compilation-related performance reasons.

**Comment by ptosco** [comment]
Release dates are reported here:
https://github.com/rdkit/rdkit/milestones
`conda` packages are published shortly thereafter.
You did not provide a reproducible example causing verbose `rdDetermineBonds` output, so I am unable to test your use case.
However, you can catch the output of `Mol.Debug()`, if that's where the verbose output you are observing comes from, as follows:
```
from contextlib import redirect_stderr
from io import StringIO
from rdkit import Chem

mol = Chem.MolFromSmiles("n1ccccc1")
with StringIO() as buf:
    with redirect_stderr(buf):
        mol.Debug()
        res = buf.getvalue()
print(res)
```
if you do not care about `buf.getvalue()`, just ignore it - I printed it above just to show how `stderr` capture can be achieved.

**Comment by bp-kelley** [comment]
@ptosco the bug is fixed, there was a call to DebugMol in determine bonds, see #7690

**Comment by bp-kelley** [comment]
Resolved in milestone https://github.com/rdkit/rdkit/milestone/96

**Comment by amorehead** [comment]
Thanks, @bp-kelley! I take it I'll be able to use this bug fix with the `2024_03_6` `conda` version of RDKit once it's released in early September?

**Comment by ptosco** [comment]
@amorehead Yes. What I was suggesting above is that you should already be able to get rid of the undesired logging with the current RDKit version by wrapping the `DetermineBonds` call as follows:
```
from contextlib import redirect_stderr
from io import StringIO

with StringIO() as buf:
    with redirect_stderr(buf):
        rdDetermineBonds.DetermineBonds(mol, charge=charge)
```

**Comment by bp-kelley** [comment]
Yes.  If you have a more concrete example of a call that's failing, I can test it for you, but this was the only debug call I found.

**Comment by amorehead** [comment]
@ptosco, sadly, it doesn't look like that worked. I'm still seeing the logs coming from `DetermineMonds()` even when redirecting the buffer manually.

**Comment by ptosco** [comment]
@amorehead If you can share the SMILES of a molecule that raises those logs I can take a look.

**Comment by amorehead** [comment]
@ptosco, an example of a SMILES string I'm giving to `rdDetermineBonds.DetermineBonds()` via
```python
charge = Chem.GetFormalCharge(mol)
            with StringIO() as buf:
                with redirect_stderr(buf):
                    # redirect RDKit's stderr to a buffer to suppress warnings
                    rdDetermineBonds.DetermineBonds(mol, charge=charge)
```
is as follows.

SMILES: `[C-3][C+]([C-3])[C+]([N-2])[C][O-]`

**Comment by ptosco** [comment]
@amorehead That SMILES yields an exception, not a warning, so it is expected that it will not be suppressed:
```
mol = Chem.MolFromSmiles("[C-3][C+]([C-3])[C+]([N-2])[C][O-]")
charge = Chem.GetFormalCharge(mol)
with StringIO() as buf:
    with redirect_stderr(buf):
        # redirect RDKit's stderr to a buffer to suppress warnings
        rdDetermineBonds.DetermineBonds(mol, charge=charge)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[70], line 5
      2 with StringIO() as buf:
      3     with redirect_stderr(buf):
      4         # redirect RDKit's stderr to a buffer to suppress warnings
----> 5         rdDetermineBonds.DetermineBonds(mol, charge=charge)

ValueError: Final molecular charge (-7) does not match input (-2); could not find valid bond ordering
```

**Comment by amorehead** [comment]
@ptosco, I noticed that if I re-attempt bond determination with a second iteration, that often yields no exception for the same input `Chem.Mol`. For reference, here is the full function I am calling to see these exceptions/logs:

```python
def create_mol_from_atom_positions_and_types(
    name: str,
    atom_positions: np.ndarray,
    element_types: List[str],
    missing_atom_indices: Set[int],
    num_bond_attempts: int = 2,
    verbose: bool = False,
) -> Mol:
    """Create an RDKit molecule from a NumPy array of atom positions and a list of their element
    types.

    :param name: The name of the molecule.
    :param atom_positions: A NumPy array of shape (num_atoms, 3) containing the 3D coordinates of
        each atom.
    :param element_types: A list of element symbols for each atom in the molecule.
    :param missing_atom_indices: A set of atom indices that are missing from the atom_positions
        array.
    :param num_bond_attempts: The number of attempts to determine the bonds in the molecule.
    :param verbose: Whether to log warnings when bond determination fails.
    :return: An RDKit molecule with the specified atom positions and element types.
    """
    if len(atom_positions) != len(element_types):
        raise ValueError("The length of atom_elements and xyz_coordinates must be the same.")

    # populate an empty editable molecule

    mol = Chem.RWMol()
    mol.SetProp("_Name", name)

    for element_type in element_types:
        atom = Chem.Atom(element_type)
        mol.AddAtom(atom)

    # set 3D coordinates

    conf = Chem.Conformer(mol.GetNumAtoms())
    for i, (x, y, z) in enumerate(atom_positions):
        conf.SetAtomPosition(i, Point3D(x, y, z))

    # add the conformer to the molecule

    mol.AddConformer(conf)
    Chem.SanitizeMol(mol)

    # block the RDKit logger

    blocker = rdBase.BlockLogs()

    # finalize molecule by inferring bonds

    determined_bonds = False
    for i in range(num_bond_attempts):
        try:
            charge = Chem.GetFormalCharge(mol)
            with StringIO() as buf:
                with redirect_stderr(buf):
                    # redirect RDKit's stderr to a buffer to suppress warnings
                    rdDetermineBonds.DetermineBonds(mol, charge=charge)
            determined_bonds = True
        except Exception as e:
            if verbose:
                logger.warning(
                    f"Failed to determine bonds for the input molecule {name} due to: {e}. "
                    f"{'Retrying once more.' if i < num_bond_attempts - 1 else 'Terminating bond assignment.'}"
                )
            continue
    if not determined_bonds:
        if verbose:
            logger.warning(
                "Failed to determine bonds in the input molecule. Skipping bond assignment."
            )

    # unblock the RDKit logger

    del blocker

    # clean up the molecule

    mol = Chem.RemoveHs(mol, sanitize=False)
    Chem.SanitizeMol(mol, catchErrors=True)

    # set a property to indicate the atom positions that are missing

    mol.SetProp("missing_atom_indices", ",".join(map(str, sorted(missing_atom_indices))))

    return mol
```

**Comment by ptosco** [comment]
That specific molecule (which has a very weird and highly unlikely charge state) does not seem to converge also after multiple iterations. Furthermore, after the first iteration all bonds are removed, so I don't think the iterative approach makes sense.

**Comment by amorehead** [comment]
@ptosco, I should clarify. The SMILES string I listed above was incorrect. It represented the output of two iterations of (failed) bond determination. The input SMILES string is rather simple:

SMILES:
```python
Chem.MolToSmiles(mol)
'C.C.C.C.C.N.O'
Chem.MolToSmiles(mol, canonical=False)
'N.C.C.C.O.C.C'  # This represents a modified amino acid residue that is being treated as a `Chem.Mol` for which to infer bonds (since sometimes arbitrary atoms can be missing for these residues)
```

**Comment by ptosco** [comment]
@amorehead Ah, sorry, I have just realized that my request for SMILES was silly - I was still focusing on `Mol.Debug()`, but `DetermineBonds()` needs 3D coordinates to infer bonds, so I will need something your starting molecule in PDB or XYZ format.

**Comment by amorehead** [comment]
@ptosco, since this function is embedded within a complex dataloading pipeline in a dedicated other repository, I can only share the inputs to the function for a single residue.

```bash
mol = create_mol_from_atom_positions_and_types(
# NOTE: for now, we construct molecules without referencing canonical
# SMILES strings, which means there are no missing molecule atoms by design
name=seq,
atom_positions=res_atom_positions[res_atom_mask],
element_types=res_atom_elements,
missing_atom_indices=set(),
verbose=verbose,
)
```
where `seq='DVA'`, `res_atom_positions[res_atom_mask]=array([[ 4.58799982, 18.20400047, 11.27400017],
       [ 5.28700018, 16.99300003, 11.63000011],
       [ 5.76599979, 16.98900032, 13.06400013],
       [ 4.34299994, 15.79500008, 11.33300018],
       [ 4.95699978, 17.36300087, 13.90799999],
       [ 4.34499979, 14.62300014, 12.31900024],
       [ 4.47200012, 15.28499985,  9.89000034]])`, and `res_atom_elements=['N', 'C', 'C', 'C', 'O', 'C', 'C']`. `verbose` also happens to be `False` here, but that doesn't seem to silence RDKit's logs. :upside_down_face:

**Comment by ptosco** [comment]
OK, I managed to reproduce the issue - I guess you refer to this message:
```
[19:18:38] Explicit valence for atom # 4 O, 3, is greater than permitted
```
That one is issued when you call `Chem.SanitizeMol(mol, catchErrors=True)`, by which time you have already called
```
    # unblock the RDKit logger

    del blocker
```
If you postpone the deletion of `blocker` until after you have called `Chem.SanitizeMol` all logging will be suppressed.

That said, I think
1. You do not need to sanitize the molecule ahead of calling `DetermineBonds`
2. Since in the absence of explicit Hs you usually have no information on formal charges from a PDB file, you can skip the charge calculation and set `allowChargedFragments=False`. By doing that you avoid that the algorithm attempts to infer bogus charges in the absence of hydrogens (and eventually fails) and can skip the iterations altogether. You can then convert the radical electrons into explicit Hs in the assumption that your PDB molecule is a neutral stable molecule.

Here's the notebook updated with my suggestions:
```
import numpy as np
from io import StringIO
from contextlib import redirect_stderr
import logging
from typing import *
from rdkit import Chem, rdBase
from rdkit.Chem import Mol
from rdkit.Chem.rdDetermineBonds import DetermineBonds
from rdkit.Geometry import Point3D
import nglview as nv

logger = logging.getLogger(__name__)

res_atom_positions = [
    [ 4.58799982, 18.20400047, 11.27400017],
    [ 5.28700018, 16.99300003, 11.63000011],
    [ 5.76599979, 16.98900032, 13.06400013],
    [ 4.34299994, 15.79500008, 11.33300018],
    [ 4.95699978, 17.36300087, 13.90799999],
    [ 4.34499979, 14.62300014, 12.31900024],
    [ 4.47200012, 15.28499985,  9.89000034]
]
res_atom_elements = ['N', 'C', 'C', 'C', 'O', 'C', 'C']

def create_mol_from_atom_positions_and_types(
    name: str,
    atom_positions: np.ndarray,
    element_types: List[str],
    missing_atom_indices: Set[int],
    num_bond_attempts: int = 2,
    verbose: bool = False,
) -> Mol:
    """Create an RDKit molecule from a NumPy array of atom positions and a list of their element
    types.

    :param name: The name of the molecule.
    :param atom_positions: A NumPy array of shape (num_atoms, 3) containing the 3D coordinates of
        each atom.
    :param element_types: A list of element symbols for each atom in the molecule.
    :param missing_atom_indices: A set of atom indices that are missing from the atom_positions
        array.
    :param num_bond_attempts: The number of attempts to determine the bonds in the molecule.
    :param verbose: Whether to log warnings when bond determination fails.
    :return: An RDKit molecule with the specified atom positions and element types.
    """
    if len(atom_positions) != len(element_types):
        raise ValueError("The length of atom_elements and xyz_coordinates must be the same.")

    # populate an empty editable molecule

    mol = Chem.RWMol()
    mol.SetProp("_Name", name)

    for element_type in element_types:
        atom = Chem.Atom(element_type)
        mol.AddAtom(atom)

    # set 3D coordinates

    conf = Chem.Conformer(mol.GetNumAtoms())
    for i, (x, y, z) in enumerate(atom_positions):
        conf.SetAtomPosition(i, Point3D(x, y, z))

    # add the conformer to the molecule

    mol.AddConformer(conf)

    # block the RDKit logger

    blocker = rdBase.BlockLogs()

    # finalize molecule by inferring bonds

    determined_bonds = False
    try:
        with StringIO() as buf:
            with redirect_stderr(buf):
                # redirect RDKit's stderr to a buffer to suppress warnings
                rdDetermineBonds.DetermineBonds(mol, allowChargedFragments=False)
        determined_bonds = True
    except Exception as e:
        pass
    if not determined_bonds:
        if verbose:
            logger.warning(
                "Failed to determine bonds in the input molecule. Skipping bond assignment."
            )

    # clean up the molecule

    mol = Chem.RemoveHs(mol, sanitize=False)
    Chem.SanitizeMol(mol, catchErrors=True)

    # unblock the RDKit logger

    del blocker

    # set a property to indicate the atom positions that are missing

    mol.SetProp("missing_atom_indices", ",".join(map(str, sorted(missing_atom_indices))))

    return mol

mol = create_mol_from_atom_positions_and_types(
    # NOTE: for now, we construct molecules without referencing canonical
    # SMILES strings, which means there are no missing molecule atoms by design
    name="DVA",
    atom_positions=res_atom_positions,
    element_types=res_atom_elements,
    missing_atom_indices=set(),
    verbose=False,
)

mol
```
![image](https://github.com/user-attachments/assets/07e9bb9c-b4d2-48c4-97d2-d5fac1e7b9bc)
```
for a in mol.GetAtoms():
    if a.GetNumRadicalElectrons():
        a.SetNumExplicitHs(a.GetNumRadicalElectrons())
        a.SetNumRadicalElectrons(0)
    Chem.SanitizeMol(mol)
mol
```
![image](https://github.com/user-attachments/assets/ce9316bf-fcf9-43b9-a11e-a6bab120f6b4)

```
view = nv.NGLWidget()
view.add_structure(nv.RdkitStructure(mol))
view
```
![image](https://github.com/user-attachments/assets/7e9416a0-8c95-44ad-8580-0e4a3c739651)

**Comment by amorehead** [comment]
@ptosco, thank you for your very detailed feedback on this function. The only remaining questions I have are, will calling `SetNumExplicitHs()` and `SetNumRadicalElectrons()` on each atom in a molecule in any way impact the molecular graph or number of atoms modeled in a molecule? And in which circumstances should one expect this `neutral stable molecule` assumption to not be maintained? It makes sense to me that residues derived from the PDB should be stable, however, this may not be the case in every situation (depending on experimental conditions).

**Comment by ptosco** [comment]
> @ptosco, thank you for your very detailed feedback on this function. The only remaining questions I have are, will calling `SetNumExplicitHs()` and `SetNumRadicalElectrons()` on each atom in a molecule in any way impact the molecular graph or number of atoms modeled in a molecule (n.b., I ask because hydrogens are not modeled by AF3)?

No. No hydrogen atoms are added to the molecular graph. The implicit/explicit hydrogen counts are simply counts set on the parent heavy atom.

> And in which circumstances should one expect this `neutral stable molecule` assumption to not be maintained? It makes sense to me that residues derived from the PDB should be stable, however, this may not be the case in every situation (depending on experimental conditions).

Most PDB structures do not have hydrogens, unless they were generated from neutron diffraction or other high resolution procedures. The absence of hydrogens prevents from drawing conclusions on the protonation state of ionizable functional groups. Tetraalkylammonium, tetraalkylphosphonium and trialkylsulfonium might be notable exceptions, as their charge does not depend on proton loss or gain, but you should be able to catch those as they should fail sanitization as neutral molecules. It should be safe to assume no radicals to be present.
Side note: for each crystal structure in the Protein Data Bank, small molecule ligands and cofactors are also separately available as 3D SDF files. The 3D SDF files have connectivity and bond order information, so it does not need to be inferred from 3D coordinates.

**Comment by amorehead** [comment]
Thanks once again @ptosco for your detailed explanation. This is very useful information for debugging the AF3 codebase. Regarding the 3D SDF files in the PDB, sadly, we cannot make use of those in this context, since AF3 performs custom atom-level filtering and cleanup of the polymer and ligand molecules in each PDB biomolecular complex. Therefore, the only format we have access to for these molecules (ligands or protein/RNA/DNA residues) is the mmCIF file format, which we then have to manually parse.
